### ISSUE TITLE
*: add script mode

### ISSUE THREAD

**Issue by dgonyeo** [appc/acbuild#23]
I think most scripts users will write to use acbuild will follow the same general formula:

- start the build
- perform some operations, if there's an error abort the build
- end the build

This formula doesn't require any of bash's complex control flow like for loops, and the resulting script must be run as root if it uses the `run` command. This will require careful inspection of any build scripts the user downloads, since blindly running shell scripts as root is terrifying.

I propose being able to write a file composed solely of comments and acbuild commands, in which acbuild will exec itself with every line that isn't a comment or empty. In the event of an error acbuild can call `abort` to clean up after itself, and this offers a little more security that an untrusted shell script.

Such a file could look like this:

```
# start the build
begin
# we're using alpine
dep add quay.io/coreos/alpine-sh
# this is an nginx container
run -- apk install nginx
# end the build
end ./mycoolapp.aci
```

And could be invoked like: `acbuild script ./build-my-cool-app.acb`

**Comment by dgonyeo** [comment]
While this is effectively inventing a DSL for acbuild, the rules by which it is executed should be rather non-magical, since it's a shorthand for a bash script.

**Comment by jonboulle** [comment]
As discussed OOB, we decided this wouldn't add much value and might be confused with Dockerfiles, so we're not going to pursue this.

**Comment by ranjib** [comment]
We can reuse the dockerfile dsl as well right ? That way we dont have to reinvent something thats already popular, as well as ease the porting process from docker to acbuild. It will be hard to port them as migrating to acbuild will require every adopter to understand how acbuild works.. and write acbuild specific features. On the other route, we can just swap docker build with acbuild and test it in our build farms.

**Comment by dgonyeo** [comment]
I'd prefer to not imitate their language, as then we're very close to guaranteeing that rkt behaves the same as docker. I'd prefer to have the language be based off of the current CLI syntax for acbuild, so it's very easy to understand how things are analogous to the bash files.

Once that's working, perhaps an automated dockerfile -> acbuild tool could be built (and it would probably include a bunch of disclaimers with it). If it turns out that such a conversion process can be made to be very reliable (although I bet I'll find some edge cases with that), maybe then it can be directly incorporated into acbuild the way docker2aci is used by rkt.

I also fully intend to support using acbuild via the CLI (aka how it currently works), if a script mode ever gets made. Being able to tie in other arbitrary command line tools to your build scripts is incredibly useful. Parsing the syntax for the DSL could literally be "take the line, append `acbuild` to the front of it, and execute it", which would make the cost of maintaining both modes of operation much lower.

**Comment by ranjib** [comment]
awesome. @dgonyeo will it be ok if I open a separate issue and start collecting feedback on what is acceptable from the current dockerfile DSL and what is not, ( I can start implementing those), we can decideif its worth the effort afterwards ? Im doing something very similar for [nut](https://github.com/pagerduty/nut), but thats more for Dockerfile -> lxc.

**Comment by dgonyeo** [comment]
I think I'd like to add the acbuild script mode before working on a dockerfile -> acbuild transpiler. I wasn't planning on focusing on this for a while, but if you're interested in seeing it move along faster I can spare some cycles once https://github.com/appc/acbuild/milestones/v0.3.0 is done.
